### PR TITLE
Adding stackable customElement decorators

### PR DESCRIPTION
**Pull request by rorticus**
**Type:** bug / feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code has been formatted with [`prettier`](https://prettier.io/) as per the [readme code style guidelines](./../#code-style)
* [x] Unit or Functional tests are included in the PR

**Description:**

In preparation for https://github.com/dojo/webpack-contrib/issues/43 we need to make the `customElement` decorator stackable.  By default, everything that inherits `WidgetBase` will get a `@customElement` decorator with sensible defaults. So,

```ts
export interface MyProps {
    prop: string;
}

class WorstWidgetEver extends WidgetBase<MyProps> { }
```

would become,

```ts
export interface MyProps {
    prop: string;
}

@customElement({
   tag: 'worst-widget-ever',
   attributes: ['prop']
})
class WorstWidgetEver extends WidgetBase<MyProps> { }
```

To customize the declaration, you could add your own custom element decorator, which would look like this,

```ts
export interface MyProps {
    prop: string;
}

@customElement({ tag: 'best-widget-ever' })
class WorstWidgetEver extends WidgetBase<MyProps> { }
```

and after transformation would look like this,

```ts
export interface MyProps {
    prop: string;
}

@customElement({ tag: 'best-widget-ever' })
@customElement({
   tag: 'worst-widget-ever',
   attributes: ['prop']
})
class WorstWidgetEver extends WidgetBase<MyProps> { }
```

This PR specifically changes the `customElement` decorator to not clobber any previously set values, but instead merge them in.  

Additionally, this PR also moves the default values out of the `customElement` decorator and into the `registerCustomElement` function. When `cli-build-widget` adds the decorator manually, it doesn't actually add the decorator, but instead adds the property to the prototype chain directly. This is done to avoid adding imports to the file, but has a downside of not being able to apply the same defaults that the decorator can. The fix of course is just to apply the defaults somewhere else (where the values are _used_ rather than where they are defined)

Doesn't resolve, but is related to https://github.com/dojo/webpack-contrib/issues/43

### CODE REVIEW

File: src/widget-core/decorators/customElement.ts [hunk 1]
```diff
@@ -37,22 +37,22 @@ export interface CustomElementConfig<P extends object = { [index: string]: any }
...
+export function customElement<P extends object = { [index: string]: any }>(config: CustomElementConfig<P>) {
+	const defaults = {
+		properties: [],
+		attributes: [],
+		events: [],
+		childType: CustomElementChildType.DOJO,
+		registryFactory: () => new Registry()
+	};
+	const { tag: tagName, ...configRest } = config;
+	const updatedConfig = { ...(tagName ? { tagName } : {}), ...configRest };
```
Comment (agubler) [inline src/widget-core/decorators/customElement.ts]: I find this really hard to follow what exactly is going on here... Perhaps it could be more verbose but end up clearer?

File: src/widget-core/decorators/customElement.ts [hunk 2]
```diff
@@ -37,22 +37,22 @@ export interface CustomElementConfig<P extends object = { [index: string]: any }
...
 	return function<T extends Constructor<any>>(target: T) {
 		target.prototype.__customElementDescriptor = {
-			tagName: tag,
-			attributes,
-			properties,
-			events,
-			childType,
-			registryFactory
+			...defaults,
+			...target.prototype.__customElementDescriptor,
```
Comment (agubler) [inline src/widget-core/decorators/customElement.ts]: Is this always an object at this point?

Comment (agubler) [reply src/widget-core/decorators/customElement.ts]: Ignore me, doesn't matter does it... spreading `{ ...undefined }` results in an empty object it seems.

File: src/widget-core/decorators/customElement.ts (hunk 1 above)
Comment (rorticus) [reply src/widget-core/decorators/customElement.ts]: @agubler I made some subtle changes here and added a comment. I mostly just removed the ternary (your comment below about undefined means this should destructure fine as an empty object if tag name isn't specified).  Anyways, hopefully this should be a little clearer now.
